File: Runtime/Plugins/iOS/ArtApps.swift

import Foundation

@MainActor
public class ArtApps {
    public static let shared = ArtApps()
    
    public private(set) var partnerId: String?
    public private(set) var appId: String?
    public private(set) var isInitialized = false
    
    private let startTime = Date()
    
    private var lastAdShowTime: Date? {
        get {
            let interval = UserDefaults.standard.double(forKey: "ArtApps_lastShowTime")
            return interval > 0 ? Date(timeIntervalSince1970: interval) : nil
        }
        set {
            UserDefaults.standard.set(newValue?.timeIntervalSince1970 ?? 0, forKey: "ArtApps_lastShowTime")
        }
    }
    
    private var serverRestrictionsUpdatedAt: Date?
    
    public var frequencyCapSeconds: TimeInterval = 90
    
    public private(set) var serverCooldownSeconds: TimeInterval?
    public private(set) var serverTtlSeconds: TimeInterval?
    
    private init() {}
    
    public func initialize(partnerId: String, appId: String, baseURL: String? = nil) {
         
         self.partnerId = partnerId
         self.appId = appId
         
         if let baseURL = baseURL {
             ArtAppsNetworkManager.shared.baseURL = baseURL
         }
        
         self.isInitialized = true
        
         print("[ArtApps] Initialized at \(startTime). PartnerID: \(partnerId)")
         print("[ArtApps] Initialized at \(startTime). AppID: \(appId)")
     }

     public var baseURL: String {
         get { ArtAppsNetworkManager.shared.baseURL }
         set { ArtAppsNetworkManager.shared.baseURL = newValue }
     }
    
    public func canShowAd() -> Bool {
        let now = Date()
 
         let effectiveCooldownSeconds = currentServerCooldownSeconds(at: now) ?? frequencyCapSeconds
         if let lastShow = lastAdShowTime, now.timeIntervalSince(lastShow) < effectiveCooldownSeconds {
             let source = currentServerCooldownSeconds(at: now) == nil ? "Freq Cap" : "Server Cooldown"
             print("[ArtApps] Blocked by \(source) (need \(effectiveCooldownSeconds)s, passed \(Int(now.timeIntervalSince(lastShow)))s)")
             return false
         }
         
         return true
     }
    
    public func updateServerRestrictions(cooldownSeconds: Int?, sessionGateSeconds: Int?, ttlSeconds: Int?) {
          serverRestrictionsUpdatedAt = Date()
          serverCooldownSeconds = cooldownSeconds.map { TimeInterval($0) }
          serverTtlSeconds = ttlSeconds.map { TimeInterval($0) }
      }
    
    public func didShowAd() {
        lastAdShowTime = Date()
    }
    
    private func currentServerCooldownSeconds(at now: Date) -> TimeInterval? {
        if checkTtl(at: now) { return nil }
        return serverCooldownSeconds
    }
    
    /// Returns true if TTL has expired and resets server restrictions.
    private func checkTtl(at now: Date) -> Bool {
        guard let updatedAt = serverRestrictionsUpdatedAt,
              let ttlSeconds = serverTtlSeconds,
              ttlSeconds > 0 else {
            return false
        }
        
        if now.timeIntervalSince(updatedAt) > ttlSeconds {
            print("[ArtApps] Server restrictions expired (TTL: \(ttlSeconds)s)")
            serverCooldownSeconds = nil
            serverTtlSeconds = nil
            serverRestrictionsUpdatedAt = nil
            return true
        }
        return false
    }
}

File: AndroidSource~/artapps-adapter/src/main/java/com/artapps/sdk/ArtApps.java
package com.artapps.sdk;

import android.content.Context;
import android.content.SharedPreferences;
import android.util.Log;

import java.util.Date;

public class ArtApps {
    private static final String TAG = "ArtApps";
    private static final String PREFS_NAME = "ArtAppsPrefs";
    private static final String KEY_LAST_SHOW_TIME = "ArtApps_lastShowTime";

    private static ArtApps instance;

    private String partnerId;
    private String appId;
    private boolean isInitialized;
    private Integer serverCooldownSeconds;
    private Integer serverTtlSeconds;
    private Date serverRestrictionsUpdatedAt;

    public long frequencyCapSeconds = 90L;

    public static synchronized ArtApps getInstance() {
        if (instance == null) {
            instance = new ArtApps();
        }
        return instance;
    }

    private ArtApps() {
    }

    public void initialize(String partnerId, String appId, String baseURL) {
        this.partnerId = partnerId;
        this.appId = appId;
        if (baseURL != null && !baseURL.isEmpty()) {
            ArtAppsNetworkManager.getInstance().setBaseURL(baseURL);
        }

        isInitialized = true;
        Log.d(TAG, "Initialized SDK 1.0.6. PartnerID: " + partnerId + ", AppID: " + appId);
    }

    public String getPartnerId() {
        return partnerId;
    }

    public String getAppId() {
        return appId;
    }

    public boolean isInitialized() {
        return isInitialized;
    }

    public boolean canShowAd(Context context) {
        long now = System.currentTimeMillis();
        long lastShowTime = getLastShowTime(context);
        long effectiveCooldown = getEffectiveCooldown(now);
        if (lastShowTime > 0L) {
            long elapsedSeconds = (now - lastShowTime) / 1000L;
            if (elapsedSeconds < effectiveCooldown) {
                String source = serverCooldownSeconds == null ? "Freq Cap" : "Server Cooldown";
                Log.d(TAG, "Blocked by " + source + " (need " + effectiveCooldown
                        + "s, passed " + elapsedSeconds + "s)");
                return false;
            }
        }
        return true;
    }

    public void didShowAd(Context context) {
        setLastShowTime(context, System.currentTimeMillis());
    }

    public void updateServerRestrictions(
            Integer cooldownSeconds,
            Integer sessionGateSeconds,
            Integer ttlSeconds
    ) {
        serverRestrictionsUpdatedAt = new Date();
        serverCooldownSeconds = cooldownSeconds;
        serverTtlSeconds = ttlSeconds;
    }

    private long getEffectiveCooldown(long nowMillis) {
        checkTtl(new Date(nowMillis));
        return serverCooldownSeconds != null ? serverCooldownSeconds : frequencyCapSeconds;
    }

    private void checkTtl(Date now) {
        if (serverRestrictionsUpdatedAt == null || serverTtlSeconds == null || serverTtlSeconds <= 0) {
            return;
        }

        long elapsed = (now.getTime() - serverRestrictionsUpdatedAt.getTime()) / 1000L;
        if (elapsed > serverTtlSeconds) {
            Log.d(TAG, "Server restrictions expired (TTL: " + serverTtlSeconds + "s)");
            serverCooldownSeconds = null;
            serverTtlSeconds = null;
            serverRestrictionsUpdatedAt = null;
        }
    }

    private long getLastShowTime(Context context) {
        if (context == null) {
            return 0L;
        }
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        return prefs.getLong(KEY_LAST_SHOW_TIME, 0L);
    }

    private void setLastShowTime(Context context, long time) {
        if (context == null) {
            return;
        }
        SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
        prefs.edit().putLong(KEY_LAST_SHOW_TIME, time).apply();
    }
}
